File: Image_Proc/color_detect.py
#===============================================================
#===============         COLOR TRACKING          ===============
#===============================================================
#====== The Color Tracking algorithm is written to      ========
#====== provide a multirotor drone real-time            ========
#====== image-processing abilities through serial       ========
#====== communication between RaspberryPi, responsible  ========
#====== for image processing and offset calculation,    ========
#====== and Arduino, serving as the primary flight      ========
#======                     controller                  ========
#===============================================================
#====== The color Detection System is written in Python ========
#====== using the OpenCV module as a base. The program  ========
#====== isolates large objects of specific color and    ========
#====== supplies information - horizontal and vertical  ========
#====== position respectful of the frame center as a    ========
#====== list of string objects through the serial port  ========
#====== to the Arduino. The third object in the list is ========
#====== the approximation of the distance dependant on  ========
#====== the area of the object that is been tracked as  ========
#======             it appears in the camera            ========
#===============================================================

import cv2
import numpy as np
import serial
import sys

serial_port = serial.Serial('/dev/ttyACM0', 9600) # Ports might be different depending on the setup being used

center_frame = (320,240)    # The (x,y) coordinates of the center of the frame with the resolution 640*480
radius_frame = (140)        # The minimum desired radius of the object being tracked
area_frame = 61575          # The desired area of the object that is being tracked
radius_frame_max = (200)    # The maximum desired radius of the object being tracked
area_frame_max = 125663     # The maximum desired area of the object being tracked

source = cv2.VideoCapture(0)
while(1):
    kernel_open = np.ones((10,10),np.uint8)     # Erosion values
    kernel_close = np.ones((30,30),np.uint8)    # Dilution values
    _, frame = source.read()                    # reads one frame at a time

    # Use this to get the resolution of the picture
    # print(frame.shape)

    # BGR to HSV conversion helps better isolate a single color
    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

    lower_blue = np.array([50,50,50])   # define range of green color in HSV
    upper_blue = np.array([70,255,255])

    # Threshold the HSV image to get only blue colors
    mask = cv2.inRange(hsv, lower_blue, upper_blue)
    opening = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel_open)
    closing = cv2.morphologyEx(opening, cv2.MORPH_CLOSE, kernel_close)

    # Bitwise-AND mask and original image
    res = cv2.bitwise_and(frame,frame, mask= closing)
    imgray = cv2.cvtColor(res, cv2.COLOR_BGR2GRAY)
    ret,thresh = cv2.threshold(imgray,1,255,0)
    contours, hierarchy = cv2.findContours(thresh,cv2.RETR_TREE,cv2.CHAIN_APPROX_SIMPLE)

    # Use this to see the array of the object
    # print(contours)

    # Two circle objects being drawn as a visual representation
    # of what the minimum and maximum desired areas are for the object being tracked
    cv2.circle(res,center_frame,radius_frame,(255,0,0),2)
    cv2.circle(res,center_frame,radius_frame_max,(0,0,255),2)

    # the following algorithm calculates the countours and the parameters needed for tracking
    # in case an object of green color and sufficient size is identified
    if len(contours) > 0:
        cnt = contours[0]
        (x,y),radius_obj = cv2.minEnclosingCircle(cnt)      # Parameters (center, radius) of the minimum circle that can enclose the tracked object
        center_obj = (int(x),int(y))                        # Center of the tracked object
        radius_obj = int(radius_obj)                        # Radius of the tracked object
        cv2.circle(res,center_obj,radius_obj,(0,255,0),-1)  # Visual representation of the minimum circular enclosure

        area_obj = ((radius_obj**2)*3.14159265359)          # Area of the minimum circular enclosure

        # Distance comparison between the calculated area and the desired area
        if area_obj > area_frame:
            if area_obj > area_frame_max:
                font = cv2.FONT_HERSHEY_SIMPLEX
                cv2.putText(res,'OOPS TOO CLOSE',(30,450), font, 2,(255,0,0),2)
                distance = "0\n" # means that the object is way too close, thus needs to move further away
            else:
                font = cv2.FONT_HERSHEY_SIMPLEX
                cv2.putText(res,'WARNING',(30,450), font, 2,(0,0,255),2)
                distance = "1\n" # means safe distance: not too close but not too far
        else:
            distance = "2\n" #means that the object is still far away from the camera, thus needs to move closer

        offset_hor = (str(int(center_obj[0]) - 320)+"\n")
        offset_ver = (str(int(center_obj[1]) - 240)+"\n")
        # The list of string objects that is sent to the Arduino.
        # The objects found are:
        # 1. Identification - used by the arduino to identify the first element
        # 2. Position in the Horizontal Plane with respect to the center of the frame
        # 3. Position in the Vertical Plane with respect to the center of the frame
        # 4. Distance approximation can only be judged as one of the three:
        #       a. 0 - too close to the camera at which the drone should back up
        #       b. 1 - ideal position at which the drone should remain hovering
        #       c. 2 - too far away - the drone should keep moving towards the object
        to_be_sent = ["31415", offset_hor, offset_ver, distance]

        for i in range(len(to_be_sent)):
            serial_port.write(to_be_sent[i])

    cv2.imshow('res',res)
    cv2.imshow('frame',frame)
    k = cv2.waitKey(5) & 0xFF
    if k == 27:
        break

cv2.destroyAllWindows()

File: Image_Proc/CPP/color_detect.cpp
/*===============================================================
=================         COLOR TRACKING          ===============
===============================================================*/

#include "opencv2/core/core.hpp"
#include "opencv2/flann/miniflann.hpp"
#include "opencv2/imgproc/imgproc.hpp"
#include "opencv2/photo/photo.hpp"
#include "opencv2/video/video.hpp"
#include "opencv2/features2d/features2d.hpp"
#include "opencv2/objdetect/objdetect.hpp"
#include "opencv2/calib3d/calib3d.hpp"
#include "opencv2/ml/ml.hpp"
#include "opencv2/highgui/highgui.hpp"
#include "opencv2/contrib/contrib.hpp"
#include "opencv2/core/core_c.h"
#include "opencv2/highgui/highgui_c.h"
#include "opencv2/imgproc/imgproc_c.h"

#include "PID.h" // PID LIBRARY

#include <iostream> // GENERAL PURPOSE LIBRARIES
#include <cmath> // GENERAL PURPOSE LIBRARIES

/*=============================================
====== Uncomment the following if you are =====
======= planning to use self-made POSIX =======
============   Serial Communication   =========
==============================================*/

// #include <termios.h>
// #include <fcntl.h>
// #include <sys/ioctl.h>
// #include <sys/stat.h>
// #include <sys/types.h>

#include <string.h>
#include <unistd.h> 
#include <stdio.h>

/*=============================================
=== In order to be able to use SerialStream ===
==== you need to install libserial locally ====
==============================================*/
#include <SerialStream.h> //POSIX and termios based library for Serial communication

using namespace cv; // USING THE NAMESPACE FOR OPENCV OBJECTS
using namespace std; // USING NAMESPACE FOR THE GENERAL PURPOSE OBJECTS
using namespace LibSerial; // USING NAMESPACE FOR THE Serial

/*================================================
=============== GENERAL DEFINITIONS ==============
================================================*/

int center_frame[2] = {320,240};  // The (x,y) coordinates of the center of the frame with the resolution 640*480
int radius_frame = 90;        	 // The minimum desired radius of the object being tracked
int area_frame = 25447;         // The desired area of the object that is being tracked
int radius_frame_max = 160;    // The maximum desired radius of the object being tracked
int area_frame_max = 80425;   // The maximum desired area of the object being tracked
int size[2] = {640,480};	 // The resolution of the camera

float PID_input_hor = 0;
float PID_output_hor = 0;
float Setpoint_hor = 320;          // The desired position of the quadcopter in centimeters
float Kp_hor = 0.135;              // Proportionality constant for the PID calculation of the Distance
float Ki_hor = 0.001;			   // Integral Constant for the PID calculation of Distance
float Kd_hor = 0.03;               // Derivative constant for the PID calculation of Distance
float Upper_Limit_hor = 10.0;      // Upper limit for the PID output of the distance correction
float Lower_Limit_hor = -10.0;     // Lower limit for the PID output of the distance correction
PID myPID(&PID_input_hor,&PID_output_hor,&Setpoint_hor,Kp_hor,Ki_hor,Kd_hor); //Initializing the PID instance

int lowH = 35;			//Low pass value for Hue
int highH = 60;			//High pass value for Hue
int lowS = 50;			//Low pass value for Saturation
int highS = 255;		//High pass value for Saturation
int lowV = 50;			//Low pass value for Value
int highV = 255;		//High pass value for Value

int largest_area = 15000;
int largest_contour_index;
int thresh = 100;
int max_thresh = 255;

int main(int argc, char* argv[]){

	SerialStream serial_port;			// Initialize Serial instance
	serial_port.Open("/dev/ttyACM0");	// Binding Serial to the Port
	serial_port.SetBaudRate( SerialStreamBuf::BAUD_9600 );	//Baud rate specification
	serial_port.SetVTime(1);	// Input Flow Control. Time to wait for data in tenths of a second
	serial_port.SetVMin(0);		// Input Flow Control. Minimum number of characters to read

	myPID.SetLimits(&Upper_Limit_hor,&Lower_Limit_hor); // Setting the upper and lower limit for the PID calculation

	VideoCapture source(0);	 // Capturng on the default port
	if (!source.isOpened())  // if not success, exit program
		{
			cout << "Did you forget to switch the camera # ???" << endl;
			return -1;
		}

	double Width = source.set(CV_CAP_PROP_FRAME_WIDTH,size[0]); //get the width of frames of the video
	double Height = source.set(CV_CAP_PROP_FRAME_HEIGHT,size[1]); //get the height of frames of the video

	while(true){
		
		myPID.Compute();	//Compute the PID values

		Mat original_frame;
		bool Success = source.read(original_frame);
		if (!Success)	//if not success, break loop
		{
			cout << "Sorry forgot how to camera.read" << endl;
			break;
		}

		/* HSV PROCESSING */
		Mat imageHSV;
		cvtColor(original_frame, imageHSV, COLOR_BGR2HSV);		// Conversion from BGR to HSV

		/* THRESHOLDING */
		Mat imageTHR;
		inRange(imageHSV, Scalar(lowH, lowS, lowV), Scalar(highH, highS, highV), imageTHR);	//Thresholding the HSV in accordance with the previously declared values

		/* OPENING EROSION/DILUTION */
		erode(imageTHR, imageTHR, getStructuringElement(MORPH_ELLIPSE, Size(8, 8)));
		dilate(imageTHR, imageTHR, getStructuringElement(MORPH_ELLIPSE, Size(8, 8)));
		dilate(imageTHR, imageTHR, getStructuringElement(MORPH_ELLIPSE, Size(8, 8))); 
		erode(imageTHR, imageTHR, getStructuringElement(MORPH_ELLIPSE, Size(8, 8)));

		/* PREPPING THE CONTOURS */
		/* BETTER KEEP THOSE REFERENCES AND POINTERS RIGHT */
		/* REALLY DOES MESS THINGS UP A LOT OF TIMES */
		vector<vector<Point> > contours;
		vector<Vec4i> hierarchy;
		Mat canny_output;

		/* CANNY EDGE DETECTION */
		Canny(imageTHR, canny_output, thresh, thresh*2, 3 ); // Detect edges using canny
		findContours(canny_output, contours, hierarchy, CV_RETR_TREE, CV_CHAIN_APPROX_SIMPLE, Point(0, 0) ); // Find contours

		vector<vector<Point>> contours_poly(contours.size());
  		vector<Point2f> center(contours.size());
  		vector<float> radius(contours.size());

		for( int i = 0; i < contours.size(); i++ )
	    {
	        minEnclosingCircle( Mat (contours[i]), center[i], radius[i] ); // Allows better estimation of the real size of the object, independent of the rotation
	        cout << int(center[i].x) << "\t" << int(center[i].y) << endl;
	        serial_port << int(center[i].x) << "\n";
	    }

	    /*==================================================================
	    ========================	  VISUALS		========================
		====================================================================
		============  Uncomment the following code in order	================
		============	to see the visual representation	================
		==================================================================*/

	    // Scalar color = Scalar(255,255,255); // The color of the drawn contour
		// Mat drawing = Mat::zeros(canny_output.size(), CV_8UC3 );
		// for( int i = 0; i< contours.size(); i++ )
		// {
		//     circle( drawing, center[i], (int)radius[i], color, 2, 8, 0 );
		// }
		
		// namedWindow("Original",CV_WINDOW_AUTOSIZE);
		// namedWindow("HSV",CV_WINDOW_AUTOSIZE);
		// namedWindow("Thresholding",CV_WINDOW_AUTOSIZE);
		// namedWindow( "Contours", CV_WINDOW_AUTOSIZE );
		// imshow( "Contours", drawing );		
		// imshow("Original", frame);
		// imshow("HSV", imageHSV);
		// imshow("Thresholding", imageTHR);

		if (waitKey(30) == 27)
		{
			break; 
		}
	}

	return 0;

}

